nexus/operation-ids.md

```diff
@@ -0,0 +1,172 @@
+
+Example invocations:
+
+```js
+Start(
+  'charge',
+  transactionId,
+  { customerId: 'someCustomerId', amount: 5000 },
+  'nexus://caller.example.com/callback?token=' + token
+);
+Cancel('charge', transactionId);
+```
+
+#### Handler addressability
+
+In some cases, the handler might require more than just a single identifier for a given operation, e.g. when an
+operation is backed by a workflow update, which requires both a workflow Id and an update Id. In order to avoid
+maintaining a handler side persistent mapping between caller and handler identifiers, the handler may choose to "leak"
+this detail to the caller, which would need to concatinate multiple identifiers for a given call. The example above
+becomes:
+
+```js
+Start(
+  'charge',
+  customerId + '/' + transactionId,
+  { amount: 5000 },
+  'nexus://caller.example.com/callback?token=' + token
+);
+Cancel('charge', customerId + '/' + transactionId);
+```
+
+> NOTE: This looks a lot like a nested REST style URL, e.g: `/customers/{customerId}/charges/{transactionId}`.
+
+> To take this approach to the extreme, `operation` and `id` could be combined into a single `path` argument, as shown
+> below:
+>
+> - `Start(path, input, callbackURL)`
+> - `Cancel(path)`
+> - `GetResult(path)`
+> - `Describe(path)`
+
+##### Mapping from client to handler identifiers
+
+To avoid leaking the identifier construction logic to the client, the handler would need a mapping of client identifiers
+to handler identifiers. Temporal can provide a strongly consistent store out of the box for this mapping.
+
+A single execution would need to map to multiple client identifiers to account for workflow starts and updates.
+
+Looking up an identifier in persistent storage added latency, and should be opted in to.
+
+### Handler generated identifiers
+
+With this approach, the handler returns a reference Id in response to a start call, which the caller should persist for
+use with the pre-defined RPCs as shown below:
+
+- `Start(operation, input, callbackURL, cancelBeforeStarted)` - handler responds inline or returns a reference Id
+  ([`*`](#cancel-before-started))
+- `Cancel(reference_id)` - cancel operation by handler provided reference Id
+- `GetResult(reference_id)` - get operation result by handler provided reference Id
+- `Describe(reference_id)` - get operation status by handler provided reference Id
+- `DeliverResult(token)` - deliver a result of an async operation using a token in the Start request's callback URL
+
+The handler _should_ either sign or qualify caller-provided reference Ids when dealing with untrusted sources and
+multi-tenant applications.
+
+In order to support stateless clients, the handler may provide an additional set of APIs to resolve client input to a
+reference Id. This may be something that is built into the protocol or implemented ad-hoc. To support this at the
+protocol level, we would need to add the following APIs:
+
+- `Resolve(operation, input) -> reference_id`
```

Comment:
I don't see how this can work generally. There's no guarantee that op/input is a unique combo. If it was we could just be hashing that and using that as the ID.

Reply:
This would assume that the input includes some form of unique identifier, agree that it is not always the case.

Reply:
It would need to include one to make the start call idempotent.